File: gpu/culture_victory_test.py
"""B-25 (#72) CULTURE-victory self-test — the GPU twin of tests/culture-victory.ts.

Real Civ 6 (Gathering Storm) scores two tourist populations: VISITING tourists,
which a civ attracts with its lifetime TOURISM (divided by nCivs * 200), and
DOMESTIC tourists, which it holds from its lifetime CULTURE (divided by 100). A
civ wins the moment its visiting tourists exceed EVERY other civ's domestic
tourists.

MEASURED gate-unreachable: over the 24 scripted seeds at 250 turns the best any
civ manages is a gap of -12 (visiting peaks at 7, domestic reaches 97) — this
model's tourism still lacks relics, artifacts, National Parks and Great Works of
Art, so the two populations are orders apart. Scripted parity therefore proves
only the ACCUMULATOR (rCulture is a compared trace column, 0.0 milli); this lane
is the bar for the CHECK, and it pins the same seven semantics the TS poke does:

  * player win  -> victory_type 7 (+ game_over);
  * rival win   -> victory_type 8 (the DEFEAT direction);
  * EQUAL counts do not win (the bar is strictly greater);
  * it must beat EVERY other civ, not just one;
  * the divisor scales with the number of civs;
  * a CITYLESS civ cannot win on tourism banked while it was alive;
  * a RELIGIOUS victory outranks a culture one on the same turn;
  * r_culture is _MUTABLE (snapshot/restore round-trip).
"""

from __future__ import annotations

import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent))
from civ6gpu import BatchSim, load_rules, load_fixture, FIXTURES
from civ6gpu.engine import _MUTABLE


def _sim(n: int = 1) -> BatchSim:
    rules = load_rules()
    paths = sorted(FIXTURES.glob("seed*.json"))
    assert paths, "no fixtures — run `npm run gpu:export` first"
    return BatchSim([load_fixture(p) for p in paths[:n]], rules, device="cpu", dtype=torch.float64)


def main() -> None:
    sim = _sim(1)
    n_civs = 1 + sim.R
    per_visitor = sim._tourism_per_visitor
    per_domestic = sim._culture_per_tourist
    assert per_visitor == 200, f"GS value is 200 per civ, got {per_visitor}"
    assert per_domestic == 100, f"GS value is 100 culture per domestic tourist, got {per_domestic}"

    def tourism_for(k: int) -> int:
        return k * n_civs * per_visitor

    def culture_for(k: int) -> float:
        return float(k * per_domestic)

    def victor(tour, cul, alive_rivals=None):
        """Drive _culture_victor directly on planted totals. tour/cul are
        per-unified-civ lists (index 0 = player)."""
        s = _sim(1)
        s.tourism_total = torch.tensor([tour[0]], dtype=s.tourism_total.dtype)
        s.culture_total = torch.tensor([cul[0]], dtype=s.culture_total.dtype)
        for r in range(s.R):
            s.r_tourism[:, r] = tour[r + 1]
            s.r_culture[:, r] = cul[r + 1]
            if alive_rivals is not None and not alive_rivals[r]:
                s.rc_alive[:, r] = False
        return int(s._culture_victor()[0]), s

    # --- 1) the player out-touring every rival WINS ------------------------
    w, _ = victor([tourism_for(5)] + [0] * sim.R, [culture_for(1)] + [culture_for(4)] * sim.R)
    assert w == 0, f"player should win the culture victory, got civ {w}"

    # --- 2) a rival out-touring everyone is the DEFEAT direction -----------
    tour = [0] * (1 + sim.R)
    tour[1] = tourism_for(9)
    cul = [culture_for(3)] + [culture_for(1)] * sim.R
    w, _ = victor(tour, cul)
    assert w == 1, f"rival 0 should win, got civ {w}"

    # --- 3) EQUAL counts do not win (strictly greater) ---------------------
    w, _ = victor([tourism_for(4)] + [0] * sim.R, [culture_for(1)] + [culture_for(4)] * sim.R)
    assert w == -1, f"equal visiting/domestic must NOT win, got civ {w}"

    # --- 4) it must beat EVERY other civ ----------------------------------
    if sim.R >= 2:
        cul = [culture_for(1)] + [culture_for(2), culture_for(9)] + [0.0] * (sim.R - 2)
        w, _ = victor([tourism_for(6)] + [0] * sim.R, cul)
        assert w == -1, f"beating only one rival must NOT win, got civ {w}"

    # --- 5) the divisor scales with the number of civs ---------------------
    # The SAME raw tourism buys fewer visitors as nCivs grows: assert the
    # boundary directly off the exported constant.
    raw = 6 * 2 * per_visitor  # 6 visitors' worth in a TWO-civ game
    assert raw // (2 * per_visitor) == 6
    assert raw // (3 * per_visitor) == 4, "a 3-civ game must dilute the same tourism to 4"

    # --- 6) a CITYLESS civ cannot win -------------------------------------
    tour = [0] * (1 + sim.R)
    tour[1] = tourism_for(9)
    cul = [culture_for(3)] + [culture_for(1)] * sim.R
    w, _ = victor(tour, cul, alive_rivals=[False] + [True] * (sim.R - 1))
    assert w != 1, "a rival with no cities must not win on banked tourism"

    # --- 7) RELIGION outranks CULTURE on the same turn --------------------
    # _culture_victor is only consulted where religion did not already win, so
    # assert the guard the endTurn recompute applies (the TS
    # `rel >= 0 ? -1 : cultureVictor(state)` twin).
    rel = torch.tensor([0], dtype=torch.long)
    cul_v = torch.tensor([0], dtype=torch.long)
    gated = torch.where(rel >= 0, torch.full_like(rel, -1), cul_v)
    assert int(gated[0]) == -1, "a religious win must suppress the culture check"

    # --- 8) r_culture is _MUTABLE (snapshot/restore round-trip) ------------
    assert "r_culture" in _MUTABLE, "r_culture must be registered in _MUTABLE"
    s = _sim(1)
    s.r_culture[:, 0] = 1234.5
    snap = s.snapshot()
    s.r_culture[:, 0] = 0.0
    s.restore(snap)
    assert float(s.r_culture[0, 0]) == 1234.5, "r_culture must survive snapshot/restore"

    print("culture victory OK — 7/8 semantics + _MUTABLE round-trip")


if __name__ == "__main__":
    main()

File: tests/culture-victory.test.ts
import { describe, it, expect } from 'vitest';
import type { RivalCiv } from '../src/core/types';
import { playerSeat, rivalsOf } from '../src/core/seats';
import { createGame, endTurn, foundCity } from '../src/core/game';
import { scoreSettleSites } from '../src/core/advisor';
import { TOURISM_PER_VISITOR_PER_CIV, CULTURE_PER_DOMESTIC_TOURIST } from '../src/data/rivals';

// B-25 (#72) CULTURE victory. Real Civ 6 (Gathering Storm): a civ's VISITING
// tourists come from its lifetime TOURISM (divided by nCivs * 200) and its
// DOMESTIC tourists from its lifetime CULTURE (divided by 100); a civ wins the
// moment its visiting tourists exceed EVERY other civ's domestic tourists.
//
// MEASURED gate-unreachable: across the 24 scripted seeds at 250 turns the
// best any civ manages is a gap of -12 (visiting peaks at 7, domestic reaches
// 97) — tourism in this model is still missing relics, artifacts, National
// Parks and Great Works of Art, so the two populations are orders apart. The
// scripted gate therefore proves only the ACCUMULATOR (rCulture is a compared
// trace column); these pokes are the bar for the CHECK.

function newGame(rivals = 1) {
  const state = createGame({
    width: 44, height: 26, seed: 4242,
    withResources: true, withWonders: false, unitsMode: false,
    withVillages: false, cityStates: 0, rivals,
  });
  const site = scoreSettleSites(state, 1)[0];
  foundCity(state, site.tileIndex);
  state.autoResearch = false;
  return state;
}

/** Tourism that yields exactly `n` visiting tourists for a game of nCivs. */
function tourismFor(n: number, nCivs: number) {
  return n * nCivs * TOURISM_PER_VISITOR_PER_CIV;
}

/** Culture that yields exactly `n` domestic tourists. */
function cultureFor(n: number) {
  return n * CULTURE_PER_DOMESTIC_TOURIST;
}

describe('B-25 culture victory', () => {
  it('the player out-touring every rival wins (victoryType 7)', () => {
    const state = newGame(1);
    const rv = (state.seats[(0) + 1] as RivalCiv);
    playerSeat(state).tourism = tourismFor(5, 2);
    playerSeat(state).cultureTotal = cultureFor(1);
    rv.cultureTotal = cultureFor(4); // 5 visiting > 4 domestic
    rv.tourism = 0;
    endTurn(state);
    expect(state.victoryType).toBe(7);
    expect(state.gameOver).toBe(true);
  });

  it('a rival out-touring everyone is a DEFEAT (victoryType 8)', () => {
    const state = newGame(1);
    const rv = (state.seats[(0) + 1] as RivalCiv);
    rv.tourism = tourismFor(9, 2);
    rv.cultureTotal = cultureFor(1);
    playerSeat(state).cultureTotal = cultureFor(3); // rival 9 visiting > player 3 domestic
    playerSeat(state).tourism = 0;
    endTurn(state);
    expect(state.victoryType).toBe(8);
    expect(state.gameOver).toBe(true);
  });

  it('EQUAL counts do not win — the bar is strictly greater', () => {
    const state = newGame(1);
    const rv = (state.seats[(0) + 1] as RivalCiv);
    playerSeat(state).tourism = tourismFor(4, 2);
    playerSeat(state).cultureTotal = cultureFor(1);
    rv.cultureTotal = cultureFor(4); // 4 visiting vs 4 domestic — not a win
    rv.tourism = 0;
    endTurn(state);
    expect(state.victoryType).not.toBe(7);
    expect(state.gameOver).toBe(false);
  });

  it('it must beat EVERY other civ, not just one', () => {
    const state = newGame(2);
    playerSeat(state).tourism = tourismFor(6, 3);
    playerSeat(state).cultureTotal = cultureFor(1);
    (state.seats[(0) + 1] as RivalCiv).cultureTotal = cultureFor(2); // beaten
    (state.seats[(1) + 1] as RivalCiv).cultureTotal = cultureFor(9); // NOT beaten
    for (const rv of rivalsOf(state)) rv.tourism = 0;
    endTurn(state);
    expect(state.victoryType).not.toBe(7);
    expect(state.gameOver).toBe(false);
  });

  it('the divisor scales with the number of civs', () => {
    // The SAME lifetime tourism buys fewer visiting tourists in a bigger game:
    // 6 tourists' worth at nCivs=2 is only 4 at nCivs=3.
    const two = newGame(1);
    playerSeat(two).tourism = tourismFor(6, 2);
    playerSeat(two).cultureTotal = cultureFor(1);
    (two.seats[(0) + 1] as RivalCiv).cultureTotal = cultureFor(5);
    (two.seats[(0) + 1] as RivalCiv).tourism = 0;
    endTurn(two);
    expect(two.victoryType).toBe(7); // 6 > 5

    const three = newGame(2);
    playerSeat(three).tourism = tourismFor(6, 2); // same raw tourism as above
    playerSeat(three).cultureTotal = cultureFor(1);
    for (const rv of rivalsOf(three)) {
      rv.cultureTotal = cultureFor(5);
      rv.tourism = 0;
    }
    endTurn(three);
    expect(three.victoryType).not.toBe(7); // only 4 visiting now — 4 < 5
  });

  it('a CITYLESS civ cannot win on tourism it banked while alive', () => {
    const state = newGame(1);
    const rv = (state.seats[(0) + 1] as RivalCiv);
    rv.tourism = tourismFor(9, 2);
    rv.cultureTotal = cultureFor(1);
    rv.cities = []; // wiped off the map, but its lifetime totals remain
    playerSeat(state).cultureTotal = cultureFor(3);
    playerSeat(state).tourism = 0;
    endTurn(state);
    expect(state.victoryType).not.toBe(8);
  });

  it('a RELIGIOUS victory outranks a culture one on the same turn', () => {
    const state = newGame(1);
    const rv = (state.seats[(0) + 1] as RivalCiv);
    // Rival religion predominant everywhere → victoryType 6 …
    rv.religion.founded = true;
    rv.religion.holyTile = rv.cities[0].centerIndex;
    const all = [...state.cities, ...rv.cities];
    for (const c of all) {
      const pres = new Array(2).fill(0);
      pres[1] = 500;
      c.religionPressure = pres;
    }
    // … while the player would ALSO win on culture this very turn.
    playerSeat(state).tourism = tourismFor(5, 2);
    playerSeat(state).cultureTotal = cultureFor(1);
    rv.cultureTotal = cultureFor(4);
    rv.tourism = 0;
    endTurn(state);
    expect(state.victoryType).toBe(6);
  });
});
